### PR TITLE
Implement flow allowing disconnection from online services when another client instance for same user is detected

### PR DESCRIPTION
**Pull request by bdach**
Prerequisite for https://github.com/ppy/osu-server-spectator/issues/80.

In action:

https://github.com/ppy/osu/assets/20418176/6ea21370-cae1-4cdc-b845-c4fc60cdb306

In short, when a user logs in on one client instance, and then on another, the previous instance is forcibly logged out. It can be restored to full online operation by logging in again, at which point any _other_ online instance will be forcibly logged out. So the goal here is to keep concurrency to 1.

The reason that this does a full logout rather than something more limited is (a) that it's easy to integrate with existing systems, and also (b) avoids weirdness. For instance, we _could_ theoretically allow the user to solo even with concurrent client instances, _were it not_ for the fact that `osu-server-spectator` - namely `SpectatorHub` - participates in the replay recording part of score submission, and cannot handle concurrency.

This currently has no testing other than manual, but I'm not sure anything more is feasible anyway. I'm also not sure of the general shape of this, which is why I'm opening as draft and refraining from adding tests until requested _and_ I am sure that there's at least an agreement that this looks how everyone would want it to look.

### CODE REVIEW

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs [hunk 1]
```diff
@@ -98,7 +98,9 @@ public virtual void PartRoom()
             if (JoinedRoom.Value == null)
                 return;
 
-            api.Queue(new PartRoomRequest(joinedRoom.Value));
+            if (api.State.Value == APIState.Online)
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:101]: Without this check is it actually feasible that this fires and causes issues?

Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:101]: Yes, I hit this once in testing.

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs [hunk 2]
```diff
@@ -88,6 +87,11 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
         /// </summary>
         public event Action? ResultsReady;
 
+        /// <summary>
+        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
+        /// </summary>
+        public event Action? Disconnecting;
```
Comment (peppy) [inline osu.Game/Online/Multiplayer/MultiplayerClient.cs:93]: Can probably move this to the interface?

```diff
diff --git a/osu.Game/Online/IStatefulUserHubClient.cs b/osu.Game/Online/IStatefulUserHubClient.cs
index 86105dd629..ec35be8ff7 100644
--- a/osu.Game/Online/IStatefulUserHubClient.cs
+++ b/osu.Game/Online/IStatefulUserHubClient.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Threading.Tasks;
 
 namespace osu.Game.Online
@@ -13,6 +14,11 @@ namespace osu.Game.Online
     /// </summary>
     public interface IStatefulUserHubClient
     {
+        /// <summary>
+        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
+        /// </summary>
+        public event Action? Disconnecting;
+
         Task DisconnectRequested();
     }
 }
diff --git a/osu.Game/Online/Multiplayer/MultiplayerClient.cs b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
index 79f46c2095..4d473a4b60 100644
--- a/osu.Game/Online/Multiplayer/MultiplayerClient.cs
+++ b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
@@ -32,6 +32,8 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
 
         public Action<Room, string>? PresentMatch { protected get; set; }
 
+        public event Action? Disconnecting;
+
         /// <summary>
         /// Invoked when any change occurs to the multiplayer room.
         /// </summary>
@@ -87,11 +89,6 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
         /// </summary>
         public event Action? ResultsReady;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
-        public event Action? Disconnecting;
-
         /// <summary>
         /// Whether the <see cref="MultiplayerClient"/> is currently connected.
         /// This is NOT thread safe and usage should be scheduled.
diff --git a/osu.Game/Online/Spectator/SpectatorClient.cs b/osu.Game/Online/Spectator/SpectatorClient.cs
index ca4ec52f4a..0cc78734df 100644
--- a/osu.Game/Online/Spectator/SpectatorClient.cs
+++ b/osu.Game/Online/Spectator/SpectatorClient.cs
@@ -70,9 +70,6 @@ public abstract partial class SpectatorClient : Component, ISpectatorClient
         /// </summary>
         public virtual event Action<int, long>? OnUserScoreProcessed;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
         public event Action? Disconnecting;
 
         /// <summary>

```

Comment (bdach) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:93]: I had it there but rider was graying it out, complaining that I ever only used it through the implementations, so I removed it... Will check again.

File: osu.Game/Online/OnlineStatusNotifier.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        [Resolved]
+        private INotificationOverlay? notificationOverlay { get; set; }
+
+        private IBindable<APIState> apiState = null!;
+        private IBindable<bool> multiplayerState = null!;
+        private IBindable<bool> spectatorState = null!;
+        private bool forcedDisconnection;
+
+        public OnlineStatusNotifier(Func<IScreen> getCurrentScreen)
```
Comment (peppy) [inline osu.Game/Online/OnlineStatusNotifier.cs:38]: I rewrote this class with a hope of simplifying things but I'm not sure if what I came up with is better or worse:

```diff
diff --git a/osu.Game/Online/IStatefulUserHubClient.cs b/osu.Game/Online/IStatefulUserHubClient.cs
index 86105dd629..ec35be8ff7 100644
--- a/osu.Game/Online/IStatefulUserHubClient.cs
+++ b/osu.Game/Online/IStatefulUserHubClient.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Threading.Tasks;
 
 namespace osu.Game.Online
@@ -13,6 +14,11 @@ namespace osu.Game.Online
     /// </summary>
     public interface IStatefulUserHubClient
     {
+        /// <summary>
+        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
+        /// </summary>
+        public event Action? Disconnecting;
+
         Task DisconnectRequested();
     }
 }
diff --git a/osu.Game/Online/Multiplayer/MultiplayerClient.cs b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
index 79f46c2095..4d473a4b60 100644
--- a/osu.Game/Online/Multiplayer/MultiplayerClient.cs
+++ b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
@@ -32,6 +32,8 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
 
         public Action<Room, string>? PresentMatch { protected get; set; }
 
+        public event Action? Disconnecting;
+
         /// <summary>
         /// Invoked when any change occurs to the multiplayer room.
         /// </summary>
@@ -87,11 +89,6 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
         /// </summary>
         public event Action? ResultsReady;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
-        public event Action? Disconnecting;
-
         /// <summary>
         /// Whether the <see cref="MultiplayerClient"/> is currently connected.
         /// This is NOT thread safe and usage should be scheduled.
diff --git a/osu.Game/Online/OnlineStatusNotifier.cs b/osu.Game/Online/OnlineStatusNotifier.cs
index 0cf672ac3c..d0a6b83681 100644
--- a/osu.Game/Online/OnlineStatusNotifier.cs
+++ b/osu.Game/Online/OnlineStatusNotifier.cs
@@ -17,6 +17,9 @@
 
 namespace osu.Game.Online
 {
+    /// <summary>
+    /// Handles various scenarios where connection is lost and we need to let the user know what and why.
+    /// </summary>
     public partial class OnlineStatusNotifier : Component
     {
         private readonly Func<IScreen> getCurrentScreen;
@@ -33,7 +36,11 @@ public partial class OnlineStatusNotifier : Component
         private IBindable<APIState> apiState = null!;
         private IBindable<bool> multiplayerState = null!;
         private IBindable<bool> spectatorState = null!;
-        private bool forcedDisconnection;
+
+        /// <summary>
+        /// This flag will be set to <c>true</c> when the user has been notified so we don't show more than one notification.
+        /// </summary>
+        private bool userNotified;
 
         public OnlineStatusNotifier(Func<IScreen> getCurrentScreen)
         {
@@ -51,59 +58,68 @@ private void load(IAPIProvider api)
             spectatorClient.Disconnecting += notifyAboutForcedDisconnection;
         }
 
-        private void notifyAboutForcedDisconnection()
-        {
-            if (forcedDisconnection)
-                return;
-
-            forcedDisconnection = true;
-            notificationOverlay?.Post(new SimpleErrorNotification
-            {
-                Icon = FontAwesome.Solid.ExclamationCircle,
-                Text = "You have been logged out on this device due to a login to your account on another device."
-            });
-        }
-
         protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            apiState.BindValueChanged(_ =>
+            apiState.BindValueChanged(state =>
             {
-                if (apiState.Value == APIState.Online)
-                    forcedDisconnection = false;
+                if (state.NewValue == APIState.Online)
+                {
+                    userNotified = false;
+                    return;
+                }
 
-                Scheduler.AddOnce(updateState);
+                if (state.NewValue == APIState.Offline)
+                {
+                    if (getCurrentScreen() is OnlinePlayScreen)
+                    {
+                        userNotified = true;
+                        notificationOverlay?.Post(new SimpleErrorNotification
+                        {
+                            Icon = FontAwesome.Solid.ExclamationCircle,
+                            Text = "Connection to API was lost. Can't continue with online play."
+                        });
+                    }
+                }
             });
-            multiplayerState.BindValueChanged(_ => Scheduler.AddOnce(updateState));
-            spectatorState.BindValueChanged(_ => Scheduler.AddOnce(updateState));
-        }
 
-        private void updateState()
-        {
-            if (forcedDisconnection)
-                return;
-
-            if (apiState.Value == APIState.Offline && getCurrentScreen() is OnlinePlayScreen)
+            multiplayerState.BindValueChanged(connected =>
             {
-                notificationOverlay?.Post(new SimpleErrorNotification
+                if (connected.NewValue)
                 {
-                    Icon = FontAwesome.Solid.ExclamationCircle,
-                    Text = "API connection was lost. Can't continue with online play."
-                });
-                return;
-            }
+                    userNotified = false;
+                    return;
+                }
 
-            if (!multiplayerClient.IsConnected.Value && multiplayerClient.Room != null)
+                if (multiplayerClient.Room != null)
+                {
+                    userNotified = true;
+                    notificationOverlay?.Post(new SimpleErrorNotification
+                    {
+                        Icon = FontAwesome.Solid.ExclamationCircle,
+                        Text = "Connection to the multiplayer server was lost. Exiting multiplayer."
+                    });
+                }
+            });
+
+            spectatorState.BindValueChanged(_ =>
             {
+                // TODO: handle spectator server failure somehow?
+            });
+        }
+
+        private void notifyAboutForcedDisconnection()
+        {
+            if (!userNotified)
+            {
+                userNotified = true;
                 notificationOverlay?.Post(new SimpleErrorNotification
                 {
                     Icon = FontAwesome.Solid.ExclamationCircle,
-                    Text = "Connection to the multiplayer server was lost. Exiting multiplayer."
+                    Text = "You have been logged out on this device due to a login to your account on another device."
                 });
             }
-
-            // TODO: handle spectator server failure somehow?
         }
 
         protected override void Dispose(bool isDisposing)
diff --git a/osu.Game/Online/Spectator/SpectatorClient.cs b/osu.Game/Online/Spectator/SpectatorClient.cs
index ca4ec52f4a..0cc78734df 100644
--- a/osu.Game/Online/Spectator/SpectatorClient.cs
+++ b/osu.Game/Online/Spectator/SpectatorClient.cs
@@ -70,9 +70,6 @@ public abstract partial class SpectatorClient : Component, ISpectatorClient
         /// </summary>
         public virtual event Action<int, long>? OnUserScoreProcessed;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
         public event Action? Disconnecting;
 
         /// <summary>

```

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:93]: It seems to hint that it's not used, but maybe fine to have there to enforce any future hubs follow standards..

File: osu.Game/Online/OnlineStatusNotifier.cs (hunk 3 above)
Comment (bdach) [reply osu.Game/Online/OnlineStatusNotifier.cs:38]: I'm not sure either, so I'm going to test this and apply it as per adage of "the reviewer is always right" if I see no issues with this.

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:93]: As per IRL discussion, the rider thing is less of an issue and the bigger issue is that apparently signalr doesn't like it and just refuses to have anything to do with any hub that has a client which declares an event. So this isn't happening.

File: osu.Game/Online/OnlineStatusNotifier.cs (hunk 3 above)
Comment (bdach) [reply osu.Game/Online/OnlineStatusNotifier.cs:45]: The patch needed further changes (not enough scheduling, as `IsConnected` changes are not update thread, also not enough checking of `userNotified`), but I made it work. See how you like the result.

File: osu.Game/Screens/OnlinePlay/Components/RoomManager.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Components/RoomManager.cs:101]: Looking at this again, I think there should be no issues with this. Server-side flows are supposed to correctly clean up the room anyway, so this request not being realised should have no impact on the server.

**Comment by peppy** [comment]
I think this is looking pretty decent.

**Review by peppy — APPROVED** [review]

**Pull request merged**